src/event.rs

```diff
@@ -634,7 +637,42 @@ pub struct KeyboardInput {
     pub modifiers: ModifiersState,
 }
 
-/// Describes an event from input method.
+/// Describes [input method](https://en.wikipedia.org/wiki/Input_method) events.
+///
+/// This is also called a "composition event".
+///
+/// Most keypresses using a latin-like keyboard layout simply generate a [`WindowEvent::ReceivedCharacter`].
+/// However, one couldn't possibly have a key for every single unicode character that the user might want to type
+/// - so the solution operating systems employ is to allow the user to type these using _a sequence of keypresses_ instead.
+///
+/// A prominent example of this is accents - many keyboard layouts allow you to first click the "accent key", and then
+/// the character you want to apply the accent to. This will generate the following event sequence:
+/// ```ignore
+/// // Press "`" key
+/// Ime::Preedit("`", Some(0), Some(0))
+/// // Press "E" key
+/// Ime::Commit("é")
+/// ```
+///
+/// Additionally, certain input devices are configured to display a candidate box that allow the user to select the
+/// desired character interactively. (To properly position this box, you must use [`Window::set_ime_position`].)
+///
+/// An example of a keyboard layout which uses candidate boxes is pinyin. On a latin keybaord the following event
+/// sequence could be obtained:
+/// ```ignore
+/// // Press "A" key
+/// Ime::Preedit("a", Some(1), Some(1))
+/// // Press "B" key
+/// Ime::Preedit("a b", Some(3), Some(3))
+/// // Press left arrow key
+/// Ime::Preedit("a b", Some(1), Some(1))
```

Comment:
I should note: This part is technically not true, currently our macOS implementation gives `Ime::Preedit("a b", Some(3), Some(3))`. But it _should_ give `Ime::Preedit("a b", Some(1), Some(1))`.

Reply:
I can ensure you that it'll work like that on Wayland at least, and that's how it should be on macOS as well? If it doesn't work like that it's likely a macOS bug, though, not really a blocking one...

Reply:
Yeah, it's a bug, should be possible to fix just has to do some UTF-16 to UTF-8 encoding that I didn't really see how to do easily.